File: 84/solution_dp.py
# DP

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n, result = len(heights), 0
        
        for i in range(n):
            if i + 1 < n and heights[i] <= heights[i + 1]:
                continue
            else:
                h = heights[i]
                for j in range(i, -1, -1):
                    h = min(h, heights[j])
                    w = i - j + 1
                    result = max(result, h * w)
                    
        return result
        

File: 84/solution_dp.java
// DP

class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int result = 0;
        
        for(int i = 0; i < n; i++) {
            if(i + 1 < n && heights[i] <= heights[i + 1]) {
                continue;
            }
            
            int h = heights[i], w;
            for(int j = i; j >= 0; j--) {
                h = Math.min(h, heights[j]);
                w = i - j + 1;
                result = Math.max(result, h * w);
            }
        }
        
        return result;
    }
}
